### DISCUSSION TITLE
Capture the language of the page request

### DISCUSSION THREAD

**Discussion by DIEGOHORVATTI** [vercel/next.js#55532] [Help] [answered]
### Summary

Motivation: I tried to access the window object and the navigation but without success, not even with if and not even doing a count and setCount to update the useEffect. I expected that this would happen due to the server request that would not let the window be manipulated by server and would only be available after mounting on the client side, but I found it strange that it still didn't work

On my system it is like this and the page there in the index should launch the user to some base language that I didn't want to be a default it could only be the default language if nothing was found does anyone have any idea what to do in this use case?

After configuring the dynamic directory [lang] I realized that the app should also have a page but it should point to the language that made the page request

To make it clearer, the tree looks like this

<img width="278" alt="image" src="https://github.com/vercel/next.js/assets/72998710/3645b49f-57a9-4acb-9aac-6a2d226f20ef">

So I would have to have something like this

<img width="739" alt="image" src="https://github.com/vercel/next.js/assets/72998710/68065915-bb54-4693-a7f1-818fddcd368b">

The whole tree

<img width="275" alt="image" src="https://github.com/vercel/next.js/assets/72998710/1148349e-f36a-4e3a-bd18-fee51b8ef80c">

**Comment by DIEGOHORVATTI** [comment]
The @philippeauriach  asked a similar question that doesn't seem to have an answer yet  #26469

**Comment by icyJoseph** [comment]
Could you describe in a flow what is you are expecting? Like:

1. user navigates to `abc`
2. user is moved to `xyz`

But with actual urls and paths.

**Reply by DIEGOHORVATTI** [reply]
@icyJoseph I don't know how this would solve my problem, I really don't understand your point if you could explain it better, because in [lang] everything is working correctly

**Reply by DIEGOHORVATTI** [reply]
only index/3000 would be the current problem that I'm solving with a workaround using a php page

**Reply by icyJoseph** [reply]
Exactly, because when you go to the root page, there's no lang in the route.

Which does make any sense the way it is currently explained. So if you want any help or input, could you first please outline what kind of URL sequence you are looking for?

**Reply by DIEGOHORVATTI** [reply]
then php retrieves the user's IP and redirects it to 3000/[lang]

**Reply by DIEGOHORVATTI** [reply]
@icyJoseph I'm waiting for it http://localhost:3000/pt

**Reply by DIEGOHORVATTI** [reply]
I wonder if there is a way to do some work around this because if we had a function that returns some customer request data such as language and current country it would be very interesting

**Reply by icyJoseph** [reply]
So what I'm gathering here:

When you go to `/pt` you cannot see the language inside the `/[lang]/page.tsx` main component? 

What next js version do you have?

**Reply by icyJoseph** [reply]
I'm still not sure what problem you are facing, it's not clear in any of your comments.

I'm assuming you have some issues with reading lang but it's not clear what or who

**Reply by DIEGOHORVATTI** [reply]
yes, there everything works correctly at build time, I compile all languages, the only problem is that I have my domain being localhost:3000/pt, this works correctly as well as localhost:3000/en but now just think if not there is not even a language set after the path localhost/3000/??? If the user ends up here by chance, he has to be redirected to localhost/3000/[user language] that's my problem currently, it's this domain root page that has to redirect the user to his language as if he were accessing the system via first time for the root domain

**Reply by icyJoseph** [reply, marked as answer]
Aha, but for that you can use a server component, or in a client component perhaps the `navigator.language` property.

```tsx
import { headers } from "next/headers";
import { redirect } from "next/navigation";

export default function Home() {
  const locale = headers().get("accept-language");
  console.log(locale);
  redirect(`/${locale?.split(",")[0]}`);
}
```

Or as you had done in the use-effect, but you calculate lang from `navigator.language` inside the use-effect, because that's a client API:

```tsx
"use client";

import { useRouter } from "next/navigation";
import { useEffect } from "react";

export default function Home() {
  const router = useRouter();

  useEffect(() => {
    const locale = navigator.language;
    router.push(`/${locale?.split(",")[0]}`);
  }, [router]);

  return null;
}

```

**Comment by DIEGOHORVATTI** [comment]
> Aha, but for that you can use a server component, or in a client component perhaps the `navigator.language` property.
> 
> ```tsx
> import { headers } from "next/headers";
> import { redirect } from "next/navigation";
> 
> export default function Home() {
>   const locale = headers().get("accept-language");
>   console.log(locale);
>   redirect(`/${locale?.split(",")[0]}`);
> }
> ```
> 
> Or as you had done in the use-effect, but you calculate lang from `navigator.language` inside the use-effect, because that's a client API:
> 
> ```tsx
> "use client";
> 
> import { useRouter } from "next/navigation";
> import { useEffect } from "react";
> 
> export default function Home() {
>   const router = useRouter();
> 
>   useEffect(() => {
>     const locale = navigator.language;
>     router.push(`/${locale?.split(",")[0]}`);
>   }, [router]);
> 
>   return null;
> }
> ```

It worked perfectly brilliant, thank you very much partner, I had no idea these headers existed

**Reply by icyJoseph** [reply]
No worries, it is tricky! Also it's probably a good idea to do a null check, so if the locale is missing you send the user to a fallback language.
